### PR TITLE
CORDA-2833 nodeinfo signing tool

### PR DESCRIPTION
**Pull request by jamesbr3**
Standalone tool to create and sign nodeInfo files, without having to start a Corda node

* Identity key from JKS file is used to sign the NodeInfo
* Node address specified via commandline
* HSM not currently supported

### CODE REVIEW

File: experimental/nodeinfo/build.gradle [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+        )
+    }
+}
+
+processResources {
+    from file("$rootDir/config/dev/log4j2.xml")
+    from file("$rootDir/node-api/src/main/resources/certificates/cordadevcakeys.jks")
+}
+
+/*publish {
```
Comment (florian-f) [inline experimental/nodeinfo/build.gradle]: delete

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 2]
```diff
@@ -0,0 +1,201 @@
+package net.corda.nodeinfo
+
+import com.typesafe.config.Config
+import com.typesafe.config.ConfigFactory
+import com.typesafe.config.ConfigParseOptions
+//import com.typesafe.config.ConfigParseOptions
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: delete

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+    @Option(names = ["--keyStorePass"], description = ["Keystore password"])
+    private var keyStorePass: String? = null
+
+    @Option(names = ["--keyAlias"],  description = ["Alias of signing key"])
+    private var keyAlias: String? = "identity-private-key"
+
+    @Option(names = ["--keyPass"], description = ["Password of signing key"])
+    private var keyPass: String? = null
+
+    private fun getInput(prompt: String, isPassword: Boolean = false): String {
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: This is never used?

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+                SerializationFactoryImpl().apply {
+                    registerScheme(AMQPInspectorSerializationScheme)
+                },
+                AMQP_P2P_CONTEXT)
+    }
+
+    class NodeInfoAndSignedNodeInfo(val nodeInfo : NodeInfo,  val signedInfoNode: SignedNodeInfo)
+
+    fun generateNodeInfo() : NodeInfoAndSignedNodeInfo {
+        var keyStore   = X509KeyStore.fromFile(keyStorePath!!, keyStorePass!!)
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: This should be a val.

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+            print("\nUse --output to write results")
+        }
+
+        return 0
+    }
+
+    //
+    // return the Identity from the specified nodekeystore
+    //
+    fun identityFromKeyStore(keyStorePath: File, keyStorePass: String, alias: String = "identity-private-key") : Party {
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: This is unused and should be deleted.

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+    private fun initialiseSerialization() {
+        nodeSerializationEnv = SerializationEnvironment.with(
+                SerializationFactoryImpl().apply {
+                    registerScheme(AMQPInspectorSerializationScheme)
+                },
+                AMQP_P2P_CONTEXT)
+    }
+
+    class NodeInfoAndSignedNodeInfo(val nodeInfo : NodeInfo,  val signedInfoNode: SignedNodeInfo)
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: You could consider just using a `Pair<NodeInfo, SignedNodeInfo>`.

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+            return 0;
+        }
+        else {
+            require(addressList.size > 0){ "At least one --address must be specified" }
+            require(outputDirectory != null) { "The --outdir parameter must be specified" }
+            require(keyStorePath != null && keyStorePass != null && keyAlias != null && keyPass != null) { "The --keyStorePath, --keyStorePass and --keyAlias and --keyPass parameters must be specified" }
+        }
+
+
+        if(outputDirectory != null) {
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: Unnecessary check, `outputDirectory` cannot be null at this point. The else can be removed as well.

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+
+        var nodeInfo = NodeInfo(addressList, identityList, platformVersion, serial)
+        var serializedNodeInfo = nodeInfo.serialize()
+
+        var sig = signingKey.keyPair.sign(serializedNodeInfo.bytes).withoutKey()  // pure DigitalSignature
+        var sni = SignedNodeInfo(serializedNodeInfo, listOf(sig))
+
+        return NodeInfoAndSignedNodeInfo(nodeInfo, sni)
+    }
+
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: Too much vertical white space.

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt (hunk 3 above)
Comment (jamesbr3) [reply experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: no, but it was going to be; i'll change to call the func when password not specified

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt (hunk 4 above)
Comment (jamesbr3) [reply experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: fixed

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt (hunk 5 above)
Comment (jamesbr3) [reply experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: done

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt (hunk 7 above)
Comment (jamesbr3) [reply experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: fixed

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt (hunk 8 above)
Comment (jamesbr3) [reply experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: fixed

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt (hunk 6 above)
Comment (jamesbr3) [reply experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: ok, considered doing this but i find the semantics harder to understand. keeping as-is for the moment

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt (hunk 3 above)
Comment (jamesbr3) [reply experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: fixed by requesting password when none specified

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 9]
```diff
@@ -0,0 +1,180 @@
+package net.corda.nodeinfo
+
+import com.typesafe.config.Config
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: Unused imports, remove.

File: experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+import java.nio.file.StandardCopyOption
+import java.security.KeyStore
+import java.security.cert.CertificateFactory
+import java.time.Instant
+
+fun main(args: Array<String>) {
+    NodeInfoSigner().start(args)
+}
+
+fun nodeInfoIdentity(nodeInfo: NodeInfo) : Party {
```
Comment (florian-f) [inline experimental/nodeinfo/src/main/kotlin/net.corda.nodeinfo/NodeInfo.kt]: Unused

**Pull request merged**